### PR TITLE
feature: importing HPCToolkit data into Hatchet

### PR DESCRIPTION
**Pull request by draganaurosgrbic**

### CODE REVIEW

**Comment by ilumsden** [comment]
@draganagrbicllnl I'm going to review your code shortly, but, before I do, I wanted to let you know that you need to add license comments to the top of your new file. You can copy those comments from any of the other Python files.

File: hatchet/graphframe.py
Comment (ilumsden) [inline hatchet/graphframe.py:130]: It would be useful if you could add `max_depth`, `min_percentage_of_parent_time`, and `min_percentage_of_application_time` plus descriptions to the docstring

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py]: Only import `GraphFrame` from `hatchet.graphframe`. The others can be imported with the following:
```python
from hatchet.frame import Frame
from hatchet.node import Node
from hatchet.graph import Graph
```

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py:131]: Is `safe_unpack` expected to return tuples with different sizes depending on inputs? I've seen calls to `safe_unpack` that return 4-element, 2-element, and 1-element tuples.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py]: @pearce8 do we want all PMPI functions to be treated as equivalent to their corresponding MPI functions? If not, this `if` statement should be removed.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py]: We usually do not set the `_hatchet_nid` field of `Node` objects using information from the profile. @pearce8 what should we use instead here?

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py]: Should this be here on its own (i.e., with no conditional)? Above, you guarded this "offset update" line with various `if` statements. As a result, if one of those conditions are satisfied, you will be doing a double increase of the offset. Just want to make sure that's intended.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py]: Could any of this code for creating the `Node` objects and updating `_metrics_table` be abstracted so it could be reused in both `_read_cct` and `_parse_context`? It would help make the code more maintainable in the future.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py:405]: Shouldn't this (and everything in this function below this line) be less indented? Right now, all this code for creating the `DataFrame`, `Graph`, and `GraphFrame` are inside the for loop over "entry points". If I'm understanding the code correctly, that means that, if you had more than one entry point, the function would return early and skip all but the first entry point.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [inline hatchet/readers/hpctoolkit_reader_latest.py:6]: Missing license header

**Review by ilumsden — CHANGES_REQUESTED** [review]
I haven't run the code yet, so this is just a review based on reading the code.

Most of the code looks good. There are a few minor issues I've commented on, but, besides one possible bug in `_read_cct`, there's nothing major.

The biggest thing I'd want to see added to this PR is a unit test or two to make sure the reader is working as expected.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [reply hatchet/readers/hpctoolkit_reader_latest.py:131]: Dragana said on Slack that this is expected behavior

File: hatchet/graphframe.py
Comment (draganaurosgrbic) [reply hatchet/graphframe.py:130]: @ilumsden fixed

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @ilumsden fixed

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @ilumsden although there were multiple `if` statements above that were incrementing the `current_offset` variable, their increment was followed by a continue statement, which would prevent double increase of the offset. However, I updated this increment statement in my latest commit. I moved it up in the while loop, so I don't repeat its occurrence in multiple places. Check my latest commit.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @ilumsden fixed

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py:6]: @ilumsden fixed

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py:405]: @ilumsden @pearce8 this is not a bug, this was intentional. Although the HPCToolkit's format of storing the CCT in general allows storing more than one tree, I am only extracting the main tree which stores the calling context of the application - this tree is marked with the entryPoint with value 1 and is always present in the database. entryPoints are roots of trees, the database always contains the tree of the application, however in general it allows to store more than one tree. That can be used if we have partial unwinds, which are very small and do not contain a lot of meaningful information.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (pearce8) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @ilumsden what does "equivalent" mean in this context?

Comment (ilumsden) [reply hatchet/readers/hpctoolkit_reader_latest.py]: Literally, what this code does is it renames nodes that start with `PMPI_` so that they start with `MPI_` instead. That's what I meant by "equivalent".

Comment (pearce8) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @draganagrbicllnl what is the intent here?  Usually we will have MPI call in application -> PMPI hook to potentially re-direct to a tool like Caliper -> actual MPI function (then back up the stack).  What does it look like from HPC toolkit, and what do we want to show to the user?  I think nested MPI functions would be confusing to the user.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (pearce8) [reply hatchet/readers/hpctoolkit_reader_latest.py:405]: @draganagrbicllnl I agree partial unwinds are confusing/usually system calls etc. and likely unnecessary.  How does HPC Toolkit ensure that entryPoint 1 is the application?  Will this remain being the case going forward?  If so, I am ok with it.

Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py:405]: @pearce8 entryPoint field parsed from the database is just an integer number which indicates the type of the tree being extracted. Number one is reserved for the application and is always 1 (it will remain so going forward). The zero is for partial unwinds.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @pearce8 done

File: hatchet/tests/data/hpctoolkit-gamess/src/lustre/orion/csc322/proj-shared/wyphan/hpctoolkit/hpctoolkit/src/tool/hpcrun/gpu/gpu-application-thread-api.c
Comment (michaelmckinsey1) [inline hatchet/tests/data/hpctoolkit-gamess/src/lustre/orion/csc322/proj-shared/wyphan/hpctoolkit/hpctoolkit/src/tool/hpcrun/gpu/gpu-application-thread-api.c:1]: I don't think we need the source code for the unit tests, so I think we can remove the `src` directory from this PR. @ilumsden Thoughts?

File: hatchet/tests/hpctoolkit_latest.py
Comment (michaelmckinsey1) [inline hatchet/tests/hpctoolkit_latest.py]: `depth` is an attribute of a `Node` not a `Node.frame`. So it could be checked this way:
```suggestion
    assert graphframe.graph.roots[0]._depth == 0
```

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]

File: hatchet/tests/hpctoolkit_latest.py
Comment (michaelmckinsey1) [inline hatchet/tests/hpctoolkit_latest.py]:
```suggestion
    assert graphframe.graph.roots[0]._depth == 0
```

File: hatchet/tests/hpctoolkit_latest.py
Comment (michaelmckinsey1) [inline hatchet/tests/hpctoolkit_latest.py]:
```suggestion
    assert graphframe.graph.roots[0]._depth == 0
```

File: hatchet/tests/hpctoolkit_latest.py
Comment (michaelmckinsey1) [inline hatchet/tests/hpctoolkit_latest.py]:
```suggestion
    assert graphframe.graph.roots[0]._depth == 0
```

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @michaelmckinsey1 fixed, forgot to update tests after making modifications in the reader

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @michaelmckinsey1 fixed

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @michaelmckinsey1 fixed

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @michaelmckinsey1 fixed

File: hatchet/tests/__init__.py
Comment (michaelmckinsey1) [inline hatchet/tests/__init__.py:1]: Add
```
# Copyright 2017-2023 Lawrence Livermore National Security, LLC and other
# Hatchet Project Developers. See the top-level LICENSE file for details.
#
# SPDX-License-Identifier: MIT
```

File: hatchet/tests/hpctoolkit_latest.py
Comment (michaelmckinsey1) [inline hatchet/tests/hpctoolkit_latest.py:6]:
```suggestion
# Copyright 2017-2023 Lawrence Livermore National Security, LLC and other
# Hatchet Project Developers. See the top-level LICENSE file for details.
#
# SPDX-License-Identifier: MIT

from hatchet import GraphFrame
```

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]

File: hatchet/tests/__init__.py
Comment (draganaurosgrbic) [reply hatchet/tests/__init__.py:1]: @michaelmckinsey1 done

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py:6]: @michaelmckinsey1 done

File: hatchet/tests/data/hpctoolkit-gamess/src/lustre/orion/csc322/proj-shared/wyphan/hpctoolkit/hpctoolkit/src/tool/hpcrun/gpu/gpu-application-thread-api.c
Comment (ilumsden) [reply hatchet/tests/data/hpctoolkit-gamess/src/lustre/orion/csc322/proj-shared/wyphan/hpctoolkit/hpctoolkit/src/tool/hpcrun/gpu/gpu-application-thread-api.c:1]: @draganagrbicllnl correct me if I'm wrong about this, but I believe that information is collected by HPCToolkit as part of the profile generation process. I don't know for sure if it is required for reading the performance data, but I believe it is a component of the profile. So, if all that is true, I think we should keep it for the sake of completeness

Comment (draganaurosgrbic) [reply hatchet/tests/data/hpctoolkit-gamess/src/lustre/orion/csc322/proj-shared/wyphan/hpctoolkit/hpctoolkit/src/tool/hpcrun/gpu/gpu-application-thread-api.c:1]: @michaelmckinsey1 @ilumsden I removed the `src` directory

File: .gitignore
Comment (ilumsden) [inline .gitignore]: Why are these in `.gitignore`? If they're local development files, that's file, but please remove these lines before we merge.

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: Why is the name of the node "1"? Usually, this would be the name of the function or region that the node represents. Since the type field is "entry", is this some HPCToolkit-specific thing where there's guaranteed to be some "virtual" root node?

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: @pearce8 I know that, in the Caliper reader, we have a mapping of certain timing metrics to the names `"time"` and `"time (inc)"`. Do we want to do something similar here?

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: Why are you testing for `>=` here?

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [reply hatchet/tests/hpctoolkit_latest.py]: Similarly, we normally use the `(inc)` metric suffix for inclusive metrics and no suffix for exclusive metrics (at least those that aren't derived by Hatchet). @pearce8 should we continue using that convention here?

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: It would be useful to add another set of `assert` statements to this test that confirms that all the nodes have a depth of <= 10. That would look something like this:
```python
for node in graphframe.graph.traverse():
    assert node._depth <= 10
```

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: Similar to the comment about confirming that the depth condition is satisfied, you should add something to this test to confirm that all the nodes take at least 1% of total application time.

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [inline hatchet/tests/hpctoolkit_latest.py]: Again, you should add something to this test to confirm that all the nodes take at least 1% of the parent time.

**Review by ilumsden — CHANGES_REQUESTED** [review]
Look good. I've only got a few more changes that should be made. This time, they're mostly related to the unit tests. The main code for the reader is pretty much good to go.

However, you need to fix formatting. The CI will not run unit tests unless the code satisfies the `black` and `flake8` checks, and we won't be able to merge until we see the unit tests passing in CI. Regarding the `flake8` checks, please rebase this PR on top of `develop`. Some changes just went in that will prevent certain failures related to `flake8`.

File: .gitignore
Comment (draganaurosgrbic) [reply .gitignore]: @ilumsden done

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (draganaurosgrbic) [reply hatchet/readers/hpctoolkit_reader_latest.py]: @ilumsden can I resolve this?

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden @pearce8 I was using `(i)` for inclusive metric mapping, and `(e)` for exclusive metric mapping. I can update this to be consistent with the Caliper convention you used. Also, for HPCToolkit we support three time metrics when measuring application - `realtime`, `cputime`, and `cycles`. I can map them to `time` for consistency.

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden done

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden done

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden done

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden for the last assertion (where I am asserting the value of `gxcopy:count (i)`) this was a mistake, since that metric stores an integer. For other cases I wasn't using `==` because of floating-point comparison. However, I updated all of my test cases that they perform `==` operator on floating-point values as well, but I rounded them to two decimals first.

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden this was a mistake. The root of the application CCT is an `entry` node with a `entryPoint` value of 1 (in general, the database can store more than one `entryPoint` for partial unwinds, they store value of 0). I was storing in the model (in the `Graph`) value of 1 for the name of that node, but when visualizing to the user, I was displaying "entry". I fixed this, so that in the model I also replace 1 with a "entry" string. I also updated tests.

**Comment by draganaurosgrbic** [comment]
@ilumsden fixed the formatting

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [reply hatchet/tests/hpctoolkit_latest.py]: @draganagrbicllnl I'd hold off on the `time` mapping for now until we hear back from @pearce8. As for the `(i)` and `(e)` part, it would probably be best to change them to `(inc)` and no suffix respectively. That convention isn't actually from Caliper. It's a Hatchet convention.

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [reply hatchet/tests/hpctoolkit_latest.py]: Sounds good. Resolving

File: hatchet/tests/hpctoolkit_latest.py
Comment (ilumsden) [reply hatchet/tests/hpctoolkit_latest.py]: Sounds good. We can keep the floating point comparison as `==` for now and revert back to `>=` if we start running into issues.

File: hatchet/readers/hpctoolkit_reader_latest.py
Comment (ilumsden) [reply hatchet/readers/hpctoolkit_reader_latest.py]: Addressed in Slack

File: hatchet/tests/hpctoolkit_latest.py
Comment (draganaurosgrbic) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden done

**Review by ilumsden — APPROVED** [review]
Now that CI is passing, I'll approve this.

Before merge, I'd appreciate your feedback on the "time metric" comment from my last review @pearce8.

**Comment by ilumsden** [comment]
Note that this PR now depends on #131

File: hatchet/tests/hpctoolkit_latest.py
Comment (pearce8) [reply hatchet/tests/hpctoolkit_latest.py]: @ilumsden @draganagrbicllnl Yes, lets go with (inc) and no suffix to match the Caliper convention.

**Review by pearce8 — APPROVED** [review]

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]
@draganagrbicllnl Can you rebase on hatchet develop?

**Comment by draganaurosgrbic** [comment]
@michaelmckinsey1 rebased on develop

**Review by michaelmckinsey1 — APPROVED** [review]

**Pull request merged**
